### ISSUE TITLE
Training the model based on a different GAN + architectural changes

### ISSUE THREAD

**Issue by rlleshi** [hanweikung/face_anon_simple#5]
First of all, great work. The idea of using a single loss is very neat. 

I have some observations and questions for you:

(1) I was training the model (with the params that you've already defined) based on another GAN swapping model (simswap) to generate the anonymized face in the data row triples. 

Since you mentioned curriculum learning in your paper (w.r.t using the anonymized face as ground truth), I did the following:
- 0 - 20K iters train with an 80/20 mix of anonymized face as ground truth (80%) and real face as conditioning image (20%)
- 20K - 80K iters train with a 50/50 mix
- 80K - 400K iters train solely on anonymized face as a conditioning image and real images as ground truths

The entire dataset was about 80K triples. 10K of these triples have the anonymized face as ground truth, the rest have it as as conditioning. 

What I noticed is the ID transfer was slowly collapsing the more I trained the model. 

Until 80K iters the ID transfer is very strong, presumably because anonymized images are still being used as ground truths. But after that, the ID error and top 1 & top 5 ID accuracy start dropping. I evaluated checkpoints from iter 240K, 320K and 400K to notice this trend. 

Do you have any ideas why this is happening and how one could ameliorate this trend? 

(2) New architectures

Below is just some ideas I wanted to explore but since training takes a long time on my GPU, I was wondering if you could help me validate them.

- Using `laion/CLIP-ViT-H-14-laion2B-s32B-b79K` instead of `clip-vit-large-patch14` for identity transfer. Motivation: bigger training dataset & better model overall.
- Using SD 2.1 Turbo (stabilityai/sd-turbo) and XL turbo (stabilityai/sdxl-turbo) as the Unet. The former can be trivially integrated since its the selfsame architecture. For the latter, given some small architectural changes, I guess we'd have to update the referencenet implementation. However, given that the referencenet model is just used for identity and attribute transfer, and that's working rather well from your results, I was thinking to keep them as they are, so based on SD 2.1 and just update the main UNet to whatever SD model. Do think this is worth trying or best to update the referencenets as well?

Thanks a lot for taking the time!

**Comment by hanweikung** [comment]
Hi, thank you for your interest and for sharing your experience.

Regarding the ID transfer degradation you observed after extended training, I didn't have this issue during my experiments, so I don't have a clear answer. However, one thing you could try is training the entire UNet instead of just the transformer blocks. The training script in the repo originally only optimized the transformers, but I've just pushed a change that enables full UNet training. Feel free to pull the latest version and give it a try. It does require more GPU memory, but in my experience, training the full UNet yields slightly better image quality. (For our paper, we trained the entire UNet.)

Also, to share some context from my training process, I started with a 50/50 mix of synthetic and real images as conditioning images and trained until 360K steps. Then I continued training using only synthesized faces as conditioning images until 435K steps.

I also occasionally see minor facial artifacts like these (left: input, right: output):
skin
![Image](https://github.com/user-attachments/assets/440bb3c4-d50c-48ed-9767-99daefcaec51)
eyes
![Image](https://github.com/user-attachments/assets/103dd676-fbf1-4a13-bcf1-a2020ef357e7)
If what you're seeing is similar, it might be okay. Your dataset size seems fine too.

As for architecture changes:
-  I haven’t tried `laion/CLIP-ViT-H-14-laion2B-s32B-b79K`, but if it's a better-performing model and easy to integrate, it's certainly worth experimenting with.
- On the UNet side: I haven’t worked with SD 2.1 Turbo or SDXL Turbo either. I suggest starting with whichever option is easier. My understanding is that SD 2.1 Turbo prioritizes inference speed and is optimized for 512x512 resolution. SD 2.1, on the other hand, can generate slightly higher resolution of 768x768. So it depends on your goal---if speed is your main concern, Turbo would be worth a try. If you're optimizing for image quality, SD 2.1 trained on 768x768 images could be a better path. (For our paper, we used 512x512 due to memory constraints at the time.)

Hope this helps, and feel free to let me know if you have new observations.

**Comment by rlleshi** [comment]
Hey, thanks for the detailed response. It clears up a lot. 

(1) Regarding the ID transfer, I guess I had interpreted it from the paper that we should only keep the 50/50 mix for the initial stages of the training, therefore I kept it until 20%, while you've kept it until 80%. This clarifies the issue of ID transfer loss.

Thank you for the tip of training the entire UNet instead of just some of its attention layers, will definitely try that out. 

(2) Regarding facial artifacts, yes, that's correct, I get that as well. Another thing that I've observed is that (sometimes) too much of the source face is transferred to the output. Sometimes things like earrings, other times birthmarks, or other facial properties of the source face that normally should not be transferred. I didn't observe this behavior with your model though. Could this also be due to training only the attention layers? 

Also, could it be worth adding an attribute loss to the training? E.g., the weak feature matching loss from simswap (but only for some of the steps considering that it would be an expensive operation).

(3) Regarding the UNet. If the architecture of the UNet deviates from the 2.1 that you developed this codebase with, do you suggest also adapting the referencenet implementation such that it fits the new SD architecture, or can we rather simply adapt the fusion process of the new UNnet and the referencenets but keep the referencenets based on SD 2.1? 

Thanks again for your observations!

**Comment by hanweikung** [comment]
Regarding (2), that issue sounds more related to the training data than the training setup. In my experience, training only the attention layers mainly impacts image quality (outputs tend to be softer or less detailed) but doesn’t cause the transfer of unintended features like earrings.

For our dataset, we used the [inswapper](https://github.com/deepinsight/insightface/blob/master/examples/in_swapper/README.md) model to generate face swaps. But [FaceFusion](https://github.com/facefusion/facefusion) offers several swap models, including inswapper and simswap, so it might be worth comparing their outputs. You could also consider mixing data generated by different models to reduce the risk of overfitting to artifacts from a single source.

Regarding attribute loss, while it might help, it could be computationally expensive as you noted. I’d prioritize improving the quality of the training data, which may already mitigate much of the unwanted feature transfer.

For (3), if you're replacing the UNet with one from a different SD architecture like SDXL Turbo, then yes, I’d suggest adapting the referencenet as well. The original advantage of using matching architectures for the UNet and referencenet is that they operate within the same feature space and can share weights. If the UNet and referencenet structure diverges from each other, this design assumption no longer holds, and the advantage may be lost.

**Comment by rlleshi** [comment]
Check out these two comparisons:

(1) The GAN swapped baseline
![Image](https://github.com/user-attachments/assets/e8f453b8-6271-499b-a929-c58efa9c935c)

(2) The diffusion swapped baseline
![Image](https://github.com/user-attachments/assets/033c36c3-4a4b-4063-8cf9-aea0a43b4b03)

Compare the lip, (subtle) birthmark around the nose, and finally the noise transfer. There's some influence from the source image and this is not coming from the GAN baseline. If anything, the eye influences of the GAN output are not being carried over to the diffusion model.

**Comment by hanweikung** [comment]
Yes, while our model learns face-swapping behavior from the GAN model and can mimic that behavior to a certain extent, it doesn't necessarily reproduce identical outputs. Based on our quantitative evaluations, GAN models that are specifically designed for face swapping typically preserve source identity similarity more effectively than our model.
